src/backend/distributed/executor/insert_select_executor.c

```diff
@@ -0,0 +1,150 @@
+/*-------------------------------------------------------------------------
+ *
+ * insert_select_executor.c
+ *
+ * Executor logic for INSERT..SELECT.
+ *
+ * Copyright (c) 2017, Citus Data, Inc.
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "distributed/insert_select_executor.h"
+#include "distributed/insert_select_planner.h"
+#include "distributed/multi_copy.h"
+#include "distributed/multi_executor.h"
+#include "distributed/multi_physical_planner.h"
+#include "distributed/multi_planner.h"
+#include "distributed/transaction_management.h"
+#include "executor/executor.h"
+#include "nodes/execnodes.h"
+#include "nodes/makefuncs.h"
+#include "nodes/nodeFuncs.h"
+#include "nodes/parsenodes.h"
+#include "nodes/plannodes.h"
+#include "parser/parse_coerce.h"
+#include "parser/parsetree.h"
+#include "tcop/pquery.h"
+#include "tcop/tcopprot.h"
+#include "utils/lsyscache.h"
+#include "utils/portal.h"
+#include "utils/snapmgr.h"
+
+
+static void ExecuteSelectIntoRelation(Oid targetRelationId, List *insertTargetList,
+									  Query *selectQuery, EState *executorState);
+static void ExecuteIntoDestReceiver(Query *query, ParamListInfo params,
+									DestReceiver *dest);
+
+
+/*
+ * CoordinatorInsertSelectExecScan executes an INSERT INTO distributed_table
+ * SELECT .. query by setting up a DestReceiver that copies tuples into the
+ * distributed table and then executing the SELECT query using that DestReceiver
+ * as the tuple destination.
+ */
+TupleTableSlot *
+CoordinatorInsertSelectExecScan(CustomScanState *node)
+{
+	CitusScanState *scanState = (CitusScanState *) node;
+	TupleTableSlot *resultSlot = NULL;
+
+	if (!scanState->finishedRemoteScan)
+	{
+		EState *executorState = scanState->customScanState.ss.ps.state;
+		MultiPlan *multiPlan = scanState->multiPlan;
+		Query *selectQuery = multiPlan->insertSelectQuery;
+		List *insertTargetList = multiPlan->insertTargetList;
+		Oid targetRelationId = multiPlan->targetRelationId;
+
+		ereport(DEBUG1, (errmsg("Collecting INSERT ... SELECT results on coordinator")));
+
+		ExecuteSelectIntoRelation(targetRelationId, insertTargetList, selectQuery,
+								  executorState);
+
+		scanState->finishedRemoteScan = true;
+	}
+
+	resultSlot = ReturnTupleFromTuplestore(scanState);
+
+	return resultSlot;
+}
+
+
+/*
+ * ExecuteSelectIntoRelation executes given SELECT query and inserts the
+ * results into the target relation, which is assumed to be a distributed
+ * table.
+ */
+static void
+ExecuteSelectIntoRelation(Oid targetRelationId, List *insertTargetList,
+						  Query *selectQuery, EState *executorState)
+{
+	ParamListInfo paramListInfo = executorState->es_param_list_info;
+
+	ListCell *insertTargetCell = NULL;
+	List *columnNameList = NIL;
+	bool stopOnFailure = false;
```

Comment:
I think we need to error on inserts to reference table otherwise replicas may diverge (or it marks one of the replicas of reference table as `INVALID`):
```C
	partitionMethod = PartitionMethod(tableId);
	if (partitionMethod == DISTRIBUTE_BY_NONE)
	{
		stopOnFailure = true;
	}
```